### DISCUSSION TITLE
Passing class instance through tasks

### DISCUSSION THREAD

**Discussion by pargass31** [apache/airflow#30736] [General]
Hello all,
I have a class for ETL processing (with subclasses) and I'd like to use is in Airflow Task.
For example I try to build something like that:

```
@task
def create_etl_instance(json_data):
    if json_data['type'] == 1:
        etl_instance = CatEtlType1()
    else:
        etl_instance = CatEtlType2()
    return etl_instance

@task
def extract_from_source(etl_instance):
    etl_instance.extract_from_source()
    return etl_instance

@task
def load_in_target(etl_instance):
    etl_instance.load_in_target()
    return etl_instance
```

And a DAG with:

```
json_data = read_json("xxxx")
cat_etl = create_etl_instance(json_data)
cat_etl = extract_from_source(cat_etl)
cat_etl = load_in_target(cat_etl)
```

But I have the following error:
"Object of type CatEtl is not JSON serializable"

Does someone see a work around? I really want to use the class because the methods (extract_from_source/load_in_target) are different regarding the subclass.

Thanks in advance.

**Comment by potiuk** [comment]
Make sure your class IS json serializable. What gets retiurned by a task is stored in XCom DB as serialized value so what you return must be serializable.  The Tasks might be executed at different machines so they need to be serialized/deserialized ot be passed between the tasks. You could likely write custom xcom backend (https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/xcoms.html#custom-xcom-backends) to serialize your class and deserialize it back if you really want but you won't get around with serializing your class and deserializing it.

**Reply by pargass31** [reply]
Thanks for your answer. If I make my class json serializable I think I cannot use methods (or maybe I'm wrong?).
For example:

```
class SimpleClass:
    def __init__(self, x: int, y: int):
        self.x = x
        self.y = y
    def add(self):
        return self.x + self.y
    def specific_operation(self):
        pass

class SimpleSubClass1(SimpleClass):
    def specific_operation(self):
        return self.x * self.y

class SimpleSubClass2(SimpleClass):
    def specific_operation(self):
        return self.x ** self.y
```

```
example_1 = SimpleSubClass1(10, 2)
example_2 = SimpleSubClass2(10, 2)
print(example_1.specific_operation()) >> 20
print(example_2.specific_operation()) >> 100
```

So now if I want to create a DAG like that:

```
with DAG(
        "dag_with_class",
        start_date=datetime(2023, 1, 1),
        catchup=False
) as dag:

    @task
    def t1(input_dict: dict):
        # initialization base on input
        if input_dict['type'] == 1:  
            class_instance = SimpleSubClass1(input_dict['x'], input_dict['y'])
        else:
            class_instance = SimpleSubClass2(input_dict['x'], input_dict['y'])

        class_instance.sum_value = class_instance.add()  # common function
        class_instance.output_from_t1 = 'ok'
        return class_instance

    @task
    def t2(class_instance: SimpleClass):
        class_instance.specific_value = class_instance.specific_operation() # specific function depending on the subclass type
        class_instance.output_from_t2 = 'ok'
        return class_instance

    input_dict = {'type': 1, 'x': 10, 'y': 2}

    class_after_step_1 = t1(input_dict)
    class_after_step_2 = t2(class_after_step_1)
```

But it doesn't work because "Object of type SimpleSubClass1 is not JSON serializable;". I don't see any solution if I want to use specific methods in separate steps (I can create only one task with all the steps but I'm not going to see steps in airflow).

**Reply by potiuk** [reply]
If you make class serializable it should work. Whethere the class has methods or not, does not matter.. There is a document on that in Airflow: https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/serializers.html#serialization
